x-pack/plugins/files/public/components/upload_file/util/simple_state_subject.ts

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { merge } from 'lodash';
+import { BehaviorSubject } from 'rxjs';
+
+export class SimpleStateSubject<S extends object = {}> extends BehaviorSubject<S> {
```

Comment:
FWIW, I've recently done something similar&mdash;it avoids using `lodash` and has explicit "transitions" of state:

```ts
import {BehaviorSubject} from 'rxjs';

export type Transition<State, Args extends unknown[]> = (...args: Args) => State;
export type PureTransition<State, Args extends unknown[]> = (state: State) => Transition<State, Args>;

/**
 * Subject which defines possible states `S` and all possible transitions `T`.
 * Transitions can then be executed by `subj.trans.xyz(...)`.
 */
export class StateSubject<S, T extends Record<string, PureTransition<S, any>>> extends BehaviorSubject<S> {
  /** A collection of possible state transitions. */
  public readonly trans = {} as {
    [K in keyof T]: ReturnType<T[K]>;
  };

  constructor(state: S, transitions: T) {
    super(state);
    for (const type of Object.keys(transitions)) {
      (this.trans as any)[type] = (...args: any) => {
        const value = this.getValue();
        const next = transitions[type](value)(...args as any);
        this.next(next);
      };
    }
  }
}
```

Usage:

```ts
const file$ = new StateSubject({status: 'idle', id: ''}, {
  setStatus => state => status => ({...state, status}),
  setId => state => id => ({...state, id}),
});

file$.trans.setId('xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx');
```

Reply:
This is a really cool pattern! I think for now I am going to stick with the "simple" state subject. We could add both of these to a common package perhaps!